File: frontend/apps/ios/Pantopus/Features/Shared/ActivityFilterSheet/ActivityFilterSheet.swift
//
//  ActivityFilterSheet.swift
//  Pantopus
//
//  P5.4 — Generic activity filter. One sheet for bids / posts / tasks /
//  offers: they all share the same filter shape (status chips + sort
//  order + date range). The sheet is a thin, semantic builder on top of
//  the shared `FilterSheetShell` — it maps an `ActivityFilter` to the
//  shell's `[FilterSection]` model and parses the applied sections back
//  into an `ActivityFilter`.
//
//  Per-surface parameterisation is the `statusOptions` (and its title):
//  bids → pending / accepted / declined / completed, tasks → open /
//  in-progress / done, offers → pending / accepted / declined, posts →
//  the post intents. Sort + date range are shared across every surface.
//

import SwiftUI

// MARK: - Filter value model

/// Sort order shared by every activity surface. The value-based cases
/// only apply to surfaces that carry an amount (bids / tasks / offers);
/// posts pass the `.timeOnly` subset.
public enum ActivitySortOrder: String, Sendable, Hashable, CaseIterable {
    case newest
    case oldest
    case valueHighToLow
    case valueLowToHigh

    public var label: String {
        switch self {
        case .newest: "Newest first"
        case .oldest: "Oldest first"
        case .valueHighToLow: "Value: high to low"
        case .valueLowToHigh: "Value: low to high"
        }
    }

    /// Date-only ordering — the subset surfaces without a per-row value
    /// (My posts) expose.
    public static let timeOnly: [ActivitySortOrder] = [.newest, .oldest]

    /// Stable order in which the radio options render.
    public static let all: [ActivitySortOrder] = [.newest, .oldest, .valueHighToLow, .valueLowToHigh]

    /// Sort `items` by this order. `date` / `value` accessors project the
    /// sort key out of each element; `nil` keys sort to the end.
    public func sorted<T>(
        _ items: [T],
        date: (T) -> Date?,
        value: (T) -> Double?
    ) -> [T] {
        switch self {
        case .newest:
            items.sorted { (date($0) ?? .distantPast) > (date($1) ?? .distantPast) }
        case .oldest:
            items.sorted { (date($0) ?? .distantFuture) < (date($1) ?? .distantFuture) }
        case .valueHighToLow:
            items.sorted { (value($0) ?? -.greatestFiniteMagnitude) > (value($1) ?? -.greatestFiniteMagnitude) }
        case .valueLowToHigh:
            items.sorted { (value($0) ?? .greatestFiniteMagnitude) < (value($1) ?? .greatestFiniteMagnitude) }
        }
    }
}

/// Date-range preset shared by every activity surface. `.anytime` is the
/// no-filter position. (A bespoke "custom" from/to picker is intentionally
/// deferred — the shared shell hosts chips, not date pickers.)
public enum ActivityDateRange: String, Sendable, Hashable, CaseIterable {
    case anytime
    case today
    case week
    case month

    public var label: String {
        switch self {
        case .anytime: "Anytime"
        case .today: "Today"
        case .week: "This week"
        case .month: "This month"
        }
    }

    /// Stable chip render order.
    public static let all: [ActivityDateRange] = [.anytime, .today, .week, .month]

    /// Whether `date` falls inside this range relative to `now`.
    public func contains(_ date: Date, now: Date, calendar: Calendar = .current) -> Bool {
        switch self {
        case .anytime:
            true
        case .today:
            calendar.isDate(date, inSameDayAs: now)
        case .week:
            date >= now.addingTimeInterval(-7 * 86400)
        case .month:
            date >= now.addingTimeInterval(-30 * 86400)
        }
    }
}

/// The applied (or working) selection across the three filter dimensions.
/// The default value is the "no filter" position — `isActive` is `false`
/// so consumers can skip filtering entirely and preserve the unfiltered
/// list order.
public struct ActivityFilter: Sendable, Hashable {
    /// Selected status chip ids. Empty = all statuses.
    public var statusIds: Set<String>
    /// Selected sort order. `nil` = keep the list's natural (load) order.
    public var sort: ActivitySortOrder?
    /// Selected date-range preset. `.anytime` = no date filtering.
    public var dateRange: ActivityDateRange

    public init(
        statusIds: Set<String> = [],
        sort: ActivitySortOrder? = nil,
        dateRange: ActivityDateRange = .anytime
    ) {
        self.statusIds = statusIds
        self.sort = sort
        self.dateRange = dateRange
    }

    /// `true` when the user has narrowed or re-ordered the list in any
    /// dimension. Drives the no-op fast path in consuming view-models.
    public var isActive: Bool {
        !statusIds.isEmpty || sort != nil || dateRange != .anytime
    }

    /// Apply this filter to `items`. `statusId` projects each element's
    /// status chip id (return `nil` for "no matching chip"); `date` and
    /// `value` project the sort/date keys. Returns `items` untouched when
    /// the filter is inactive.
    public func apply<T>(
        to items: [T],
        now: Date,
        statusId: (T) -> String?,
        date: (T) -> Date?,
        value: (T) -> Double?
    ) -> [T] {
        guard isActive else { return items }
        var result = items
        if !statusIds.isEmpty {
            result = result.filter { element in
                guard let id = statusId(element) else { return false }
                return statusIds.contains(id)
            }
        }
        if dateRange != .anytime {
            result = result.filter { element in
                guard let when = date(element) else { return false }
                return dateRange.contains(when, now: now)
            }
        }
        if let sort {
            result = sort.sorted(result, date: date, value: value)
        }
        return result
    }
}

// MARK: - Section identifiers

enum ActivityFilterSection {
    static let status = "status"
    static let sort = "sort"
    static let dateRange = "dateRange"
}

// MARK: - Sheet

/// The generic activity filter sheet. Builds three sections (status /
/// sort / date range) and renders them through `FilterSheetShell`, then
/// maps the applied sections back to an `ActivityFilter`.
@MainActor
public struct ActivityFilterSheet: View {
    private let statusTitle: String
    private let statusOptions: [FilterOption]
    private let sortOptions: [ActivitySortOrder]
    private let filter: ActivityFilter
    private let onApply: @MainActor (ActivityFilter) -> Void
    private let onClose: @MainActor () -> Void

    /// - Parameters:
    ///   - statusTitle: Section header for the status chips (e.g.
    ///     `"Status"`, `"Type"`).
    ///   - statusOptions: Per-surface status chips. Pass `[]` to omit the
    ///     status section entirely.
    ///   - sortOptions: Sort orders to offer. Defaults to the full set;
    ///     surfaces without a per-row value pass `ActivitySortOrder.timeOnly`.
    ///   - filter: The currently-applied filter (seeds the working copy).
    ///   - onApply: Called with the parsed filter when the user taps Apply.
    ///   - onClose: Called when the sheet should dismiss.
    public init(
        statusTitle: String = "Status",
        statusOptions: [FilterOption],
        sortOptions: [ActivitySortOrder] = ActivitySortOrder.all,
        filter: ActivityFilter,
        onApply: @escaping @MainActor (ActivityFilter) -> Void,
        onClose: @escaping @MainActor () -> Void
    ) {
        self.statusTitle = statusTitle
        self.statusOptions = statusOptions
        self.sortOptions = sortOptions
        self.filter = filter
        self.onApply = onApply
        self.onClose = onClose
    }

    public var body: some View {
        FilterSheetShell(
            title: "Filters",
            sections: Self.sections(
                statusTitle: statusTitle,
                statusOptions: statusOptions,
                sortOptions: sortOptions,
                filter: filter
            ),
            onApply: { applied in onApply(Self.filter(from: applied)) },
            onClose: onClose
        )
        .accessibilityIdentifier("activityFilterSheet")
    }

    // MARK: - Mapping

    /// Build the shell's section list from an `ActivityFilter`.
    static func sections(
        statusTitle: String,
        statusOptions: [FilterOption],
        sortOptions: [ActivitySortOrder],
        filter: ActivityFilter
    ) -> [FilterSection] {
        var sections: [FilterSection] = []
        if !statusOptions.isEmpty {
            sections.append(
                FilterSection(
                    id: ActivityFilterSection.status,
                    title: statusTitle,
                    control: .chipGroup(options: statusOptions, selectedIds: filter.statusIds)
                )
            )
        }
        sections.append(
            FilterSection(
                id: ActivityFilterSection.sort,
                title: "Sort by",
                control: .radio(
                    options: sortOptions.map { FilterOption(id: $0.rawValue, label: $0.label) },
                    selectedId: filter.sort?.rawValue
                )
            )
        )
        sections.append(
            FilterSection(
                id: ActivityFilterSection.dateRange,
                title: "Date range",
                control: .singleChip(
                    options: ActivityDateRange.all.map { FilterOption(id: $0.rawValue, label: $0.label) },
                    selectedId: filter.dateRange.rawValue
                )
            )
        )
        return sections
    }

    /// Parse the shell's applied sections back into an `ActivityFilter`.
    static func filter(from sections: [FilterSection]) -> ActivityFilter {
        var statusIds: Set<String> = []
        var sort: ActivitySortOrder?
        var dateRange: ActivityDateRange = .anytime
        for section in sections {
            switch section.id {
            case ActivityFilterSection.status:
                if case let .chipGroup(_, ids) = section.control { statusIds = ids }
            case ActivityFilterSection.sort:
                if case let .radio(_, selectedId) = section.control {
                    sort = selectedId.flatMap(ActivitySortOrder.init(rawValue:))
                }
            case ActivityFilterSection.dateRange:
                if case let .singleChip(_, selectedId) = section.control {
                    dateRange = selectedId.flatMap(ActivityDateRange.init(rawValue:)) ?? .anytime
                }
            default:
                break
            }
        }
        return ActivityFilter(statusIds: statusIds, sort: sort, dateRange: dateRange)
    }
}

#Preview {
    ActivityFilterSheet(
        statusTitle: "Status",
        statusOptions: [
            FilterOption(id: "pending", label: "Pending"),
            FilterOption(id: "accepted", label: "Accepted"),
            FilterOption(id: "declined", label: "Declined"),
            FilterOption(id: "completed", label: "Completed")
        ],
        filter: ActivityFilter(statusIds: ["pending"], sort: .newest, dateRange: .week),
        onApply: { _ in },
        onClose: {}
    )
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/activity_filter_sheet/ActivityFilterSheet.kt
@file:Suppress("PackageNaming", "MagicNumber", "LongParameterList")
@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package app.pantopus.android.ui.screens.shared.activity_filter_sheet

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import app.pantopus.android.ui.screens.shared.filter_sheet.FilterControl
import app.pantopus.android.ui.screens.shared.filter_sheet.FilterOption
import app.pantopus.android.ui.screens.shared.filter_sheet.FilterSection
import app.pantopus.android.ui.screens.shared.filter_sheet.FilterSheetShell
import java.time.Instant
import java.time.ZoneId

/**
 * P5.4 — Generic activity filter. One sheet for bids / posts / tasks /
 * offers: they share the same filter shape (status chips + sort order +
 * date range). A thin, semantic builder on top of the shared
 * [FilterSheetShell] — mirrors the iOS `ActivityFilterSheet`.
 *
 * Per-surface parameterisation is the [statusOptions] (and its title);
 * sort + date range are shared across every surface.
 */

/** Sort order shared by every activity surface. */
enum class ActivitySortOrder(
    val id: String,
    val label: String,
) {
    Newest("newest", "Newest first"),
    Oldest("oldest", "Oldest first"),
    ValueHighToLow("valueHighToLow", "Value: high to low"),
    ValueLowToHigh("valueLowToHigh", "Value: low to high"),
    ;

    companion object {
        /** Full set — surfaces with a per-row value (bids / tasks / offers). */
        val ALL = listOf(Newest, Oldest, ValueHighToLow, ValueLowToHigh)

        /** Date-only subset — surfaces without a value (My posts). */
        val TIME_ONLY = listOf(Newest, Oldest)

        fun fromId(id: String?): ActivitySortOrder? = entries.firstOrNull { it.id == id }
    }
}

/**
 * Date-range preset shared by every activity surface. [Anytime] is the
 * no-filter position. (A bespoke "custom" picker is intentionally
 * deferred — the shared shell hosts chips, not date pickers.)
 */
enum class ActivityDateRange(
    val id: String,
    val label: String,
) {
    Anytime("anytime", "Anytime"),
    Today("today", "Today"),
    Week("week", "This week"),
    Month("month", "This month"),
    ;

    /** Whether [date] falls inside this range relative to [now]. */
    fun contains(
        date: Instant,
        now: Instant,
        zone: ZoneId = ZoneId.systemDefault(),
    ): Boolean =
        when (this) {
            Anytime -> true
            Today -> date.atZone(zone).toLocalDate() == now.atZone(zone).toLocalDate()
            Week -> !date.isBefore(now.minusSeconds(7L * 86_400))
            Month -> !date.isBefore(now.minusSeconds(30L * 86_400))
        }

    companion object {
        val ALL = listOf(Anytime, Today, Week, Month)

        fun fromId(id: String?): ActivityDateRange = entries.firstOrNull { it.id == id } ?: Anytime
    }
}

/**
 * The applied (or working) selection across the three filter dimensions.
 * The default value is the "no filter" position — [isActive] is `false`
 * so consumers can skip filtering and preserve the unfiltered order.
 */
@Immutable
data class ActivityFilter(
    /** Selected status chip ids. Empty = all statuses. */
    val statusIds: Set<String> = emptySet(),
    /** Selected sort order. `null` = keep the list's natural order. */
    val sort: ActivitySortOrder? = null,
    /** Selected date-range preset. [ActivityDateRange.Anytime] = no filter. */
    val dateRange: ActivityDateRange = ActivityDateRange.Anytime,
) {
    /** `true` when the user has narrowed or re-ordered the list. */
    val isActive: Boolean
        get() = statusIds.isNotEmpty() || sort != null || dateRange != ActivityDateRange.Anytime

    /**
     * Apply this filter to [items]. [statusId] projects each element's
     * status chip id (return `null` for "no matching chip"); [date] and
     * [value] project the sort/date keys. Returns [items] untouched when
     * the filter is inactive.
     */
    fun <T> apply(
        items: List<T>,
        now: Instant,
        statusId: (T) -> String?,
        date: (T) -> Instant?,
        value: (T) -> Double?,
    ): List<T> {
        if (!isActive) return items
        var result = items
        if (statusIds.isNotEmpty()) {
            result = result.filter { statusId(it)?.let(statusIds::contains) == true }
        }
        if (dateRange != ActivityDateRange.Anytime) {
            result = result.filter { element -> date(element)?.let { dateRange.contains(it, now) } == true }
        }
        result =
            when (sort) {
                ActivitySortOrder.Newest -> result.sortedByDescending { date(it) ?: Instant.MIN }
                ActivitySortOrder.Oldest -> result.sortedBy { date(it) ?: Instant.MAX }
                ActivitySortOrder.ValueHighToLow -> result.sortedByDescending { value(it) ?: -Double.MAX_VALUE }
                ActivitySortOrder.ValueLowToHigh -> result.sortedBy { value(it) ?: Double.MAX_VALUE }
                null -> result
            }
        return result
    }
}

/** Stable section ids shared with the iOS counterpart. */
object ActivityFilterSectionId {
    const val STATUS = "status"
    const val SORT = "sort"
    const val DATE_RANGE = "dateRange"
}

/** Build the shell's section list from an [ActivityFilter]. */
fun activityFilterSections(
    statusTitle: String,
    statusOptions: List<FilterOption>,
    sortOptions: List<ActivitySortOrder>,
    filter: ActivityFilter,
): List<FilterSection> {
    val sections = mutableListOf<FilterSection>()
    if (statusOptions.isNotEmpty()) {
        sections.add(
            FilterSection(
                id = ActivityFilterSectionId.STATUS,
                title = statusTitle,
                control = FilterControl.ChipGroup(options = statusOptions, selectedIds = filter.statusIds),
            ),
        )
    }
    sections.add(
        FilterSection(
            id = ActivityFilterSectionId.SORT,
            title = "Sort by",
            control =
                FilterControl.Radio(
                    options = sortOptions.map { FilterOption(id = it.id, label = it.label) },
                    selectedId = filter.sort?.id,
                ),
        ),
    )
    sections.add(
        FilterSection(
            id = ActivityFilterSectionId.DATE_RANGE,
            title = "Date range",
            control =
                FilterControl.SingleChip(
                    options = ActivityDateRange.ALL.map { FilterOption(id = it.id, label = it.label) },
                    selectedId = filter.dateRange.id,
                ),
        ),
    )
    return sections
}

/** Parse the shell's applied sections back into an [ActivityFilter]. */
fun activityFilterFrom(sections: List<FilterSection>): ActivityFilter {
    var statusIds: Set<String> = emptySet()
    var sort: ActivitySortOrder? = null
    var dateRange = ActivityDateRange.Anytime
    sections.forEach { section ->
        when (section.id) {
            ActivityFilterSectionId.STATUS ->
                (section.control as? FilterControl.ChipGroup)?.let { statusIds = it.selectedIds }
            ActivityFilterSectionId.SORT ->
                (section.control as? FilterControl.Radio)?.let { sort = ActivitySortOrder.fromId(it.selectedId) }
            ActivityFilterSectionId.DATE_RANGE ->
                (section.control as? FilterControl.SingleChip)?.let { dateRange = ActivityDateRange.fromId(it.selectedId) }
        }
    }
    return ActivityFilter(statusIds = statusIds, sort = sort, dateRange = dateRange)
}

/**
 * The generic activity filter sheet. Builds three sections (status /
 * sort / date range) and renders them through [FilterSheetShell], then
 * maps the applied sections back to an [ActivityFilter].
 *
 * @param statusOptions Per-surface status chips. Pass an empty list to
 *     omit the status section entirely.
 * @param sortOptions Sort orders to offer. Surfaces without a per-row
 *     value pass [ActivitySortOrder.TIME_ONLY].
 * @param filter The currently-applied filter (seeds the working copy).
 * @param onApply Called with the parsed filter when the user taps Apply.
 * @param onDismiss Called when the sheet should dismiss.
 */
@Composable
fun ActivityFilterSheet(
    statusOptions: List<FilterOption>,
    filter: ActivityFilter,
    onApply: (ActivityFilter) -> Unit,
    onDismiss: () -> Unit,
    statusTitle: String = "Status",
    sortOptions: List<ActivitySortOrder> = ActivitySortOrder.ALL,
) {
    FilterSheetShell(
        sections =
            activityFilterSections(
                statusTitle = statusTitle,
                statusOptions = statusOptions,
                sortOptions = sortOptions,
                filter = filter,
            ),
        onApply = { applied -> onApply(activityFilterFrom(applied)) },
        onDismiss = onDismiss,
        title = "Filters",
    )
}
